src/ion/liveranges.rs

```diff
@@ -0,0 +1,1201 @@
+/*
+ * The following license applies to this file, which was initially
+ * derived from the files `js/src/jit/BacktrackingAllocator.h` and
+ * `js/src/jit/BacktrackingAllocator.cpp` in Mozilla Firefox:
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ * Since the initial port, the design has been substantially evolved
+ * and optimized.
+ */
+
+//! Live-range computation.
+
+use super::{
+    CodeRange, Env, InsertMovePrio, LiveBundle, LiveBundleIndex, LiveRange, LiveRangeFlag,
+    LiveRangeIndex, LiveRangeKey, LiveRangeListEntry, LiveRangeSet, PRegData, PRegIndex, RegClass,
+    SpillSetIndex, Use, VRegData, VRegIndex, SLOT_NONE,
+};
+use crate::bitvec::BitVec;
+use crate::{
+    Allocation, Block, Function, Inst, InstPosition, Operand, OperandConstraint, OperandKind,
+    OperandPos, PReg, ProgPoint, RegAllocError, VReg,
+};
+use fxhash::FxHashSet;
+use smallvec::{smallvec, SmallVec};
+use std::collections::{HashSet, VecDeque};
+use std::convert::TryFrom;
+
+#[inline(always)]
+pub fn spill_weight_from_constraint(
+    constraint: OperandConstraint,
+    loop_depth: usize,
+    is_def: bool,
+) -> u32 {
+    // A bonus of 1000 for one loop level, 4000 for two loop levels,
```

Comment:
Imagine what happens if you have to process a loop nest that is 3 or more deep.  Because this cuts off nesting at depth 2, it can't distinguish, on a cost basis, blocks in the body of the innermost loop from blocks in the middle-level loop.  Hence you may wind up with very inaccurate cost estimates for the innermost loop blocks; yet it is exactly those that that it is most important to have a correct cost estimation for.

For BT or Ion (I forget which) I saw loop depths of up to 5 in the input.  You maybe want to consider removing this limit, and changing the fundamental cost type to be 32-bit float, so as to get rid of the dynamic range problems associated with using a `u32` (scaled in units of 1000) for costs.

Reply:
I definitely agree that a scale that goes up to higher nesting levels would be ideal here; the constraint we bump into is that we store the weight of an operand in 16 bits in `ion::Use` (it's tightly bit-packed). I am definitely OK with adding "tune the cost heuristics and possibly adjust the bitpacking as needed" as an issue to track post-initial-review, if that's acceptable, though?

Reply:
I mention this primarily because Ion has the similar logic, and I seem to
remember Ion doing stupid stuff as a result.
See https://bugzilla.mozilla.org/show_bug.cgi?id=1689328.

> we store the weight of an operand in 16 bits

Can you do poor-person's float16?  What I mean is:
* Compute the weight as a float32, as usual
* It must be zero or have sign bit 0 (positive) otherwise the system is in deep trouble :-)
* store, as your uint16, bits 30:15 of that value.  That gets you all 8 bits of the exponent and the most significant 8 bits of the mantissa, and has the property (I think!) that the total ordering on those uint16s is isomorphic to the total ordering of the original float32s.

Reply:
.. assuming no NaNs, Infs, denorms, or other mutantcy in the f32s.

Reply:
Or maybe better https://en.wikipedia.org/wiki/Bfloat16_floating-point_format

"The bfloat16 format, being a truncated IEEE 754 single-precision 32-bit float, 
allows for fast conversion to and from an IEEE 754 single-precision 32-bit float"

Then you could plausibly do arithmetic etc on the 16-bit values, if you need to 
(eg) add costs together when coalescing, etc

Reply:
Ah, yes, that's a great idea -- I hadn't considered using a low-precision float but it fits the use-case perfectly (we want high dynamic range but not necessarily a lot of precision). Done!